editor/src/components/canvas/controls/formula-bar.tsx

```diff
@@ -0,0 +1,111 @@
+import * as React from 'react'
+import * as EditorActions from '../../editor/actions/action-creators'
+import { betterReactMemo } from '../../../uuiui-deps'
+import { colorTheme, SimpleFlexRow } from '../../../uuiui'
+import { useEditorState } from '../../editor/store/store-hook'
+import { MetadataUtils } from '../../../core/model/element-metadata-utils'
+import { isRight } from '../../../core/shared/either'
+import * as TP from '../../../core/shared/template-path'
+import {
+  isJSXArbitraryBlock,
+  isJSXElement,
+  isJSXTextBlock,
+} from '../../../core/shared/element-template'
+
+export const FormulaBar = betterReactMemo('FormulaBar', () => {
+  const saveTimerRef = React.useRef<any>(null)
+  const dispatch = useEditorState((store) => store.dispatch, 'FormulaBar dispatch')
+
+  const selectedElement = useEditorState((store) => {
+    const metadata = store.editor.jsxMetadataKILLME
+    if (
+      store.editor.selectedViews.length === 1 &&
+      TP.isInstancePath(store.editor.selectedViews[0])
+    ) {
+      return MetadataUtils.getElementByInstancePathMaybe(
+        metadata.elements,
+        store.editor.selectedViews[0],
+      )
+    } else {
+      return null
+    }
+  }, 'Formula Bar selectedElement')
+
+  const [simpleText, setSimpleText] = React.useState('')
+  const [disabled, setDisabled] = React.useState(false)
+
+  React.useEffect(() => {
+    if (saveTimerRef.current != null) {
```

Comment:
What's this for as it doesn't look like it's used in this block?

Reply:
it is used for keeping the inputfield state when a user is typing while the save action is still waiting inside the timer

Reply:
it looks like if the user changes the element while the timer is still running, the formula bar is going to remain unchanged, showing the text of the previously selected element.

This is beacuse you use a ref for the timer, which is tied to the lifecycle of the FormulaBar component, so if the selected element changes, the ref is still pointing to an active timer, which we do not want to cancel if the user types more, because that would lead to a loss of the data of the last few keystrokes on the previously selected element.

I think the best solution to solve this would be to remount the FormulaBar when selection changes. Then the old ref will remain intact while the timer runs out, and we create a new ref for the newly selected element.
https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key

Reply:
fixed